cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt

```diff
@@ -771,9 +770,7 @@ class ExpressionHandler(lang: JavaLanguageFrontend) :
     private fun createImplicitThis():
         de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression {
         val base: de.fraunhofer.aisec.cpg.graph.statements.expressions.Expression
-        val thisType =
-            (frontend.scopeManager.currentFunction as? MethodDeclaration?)?.receiver?.type
-                ?: unknownType()
+        val thisType = frontend.scopeManager.currentMethod?.receiver?.type ?: unknownType()
         base = newReference("this", thisType).implicit("this")
```

Comment:
Shouldn't this be the current class instead of using the current method and the receiver? Seems more error-prone that using the current class.

Reply:
What about one of those C++ situations where a function definition is not inside the class definition? I could see that there, the receiver is still correct but the current class points to the wrong type for the this reference.

Reply:
> What about one of those C++ situations where a function definition is not inside the class definition? I could see that there, the receiver is still correct but the current class points to the wrong type for the this reference.

No, in this case, the "currentClass" is also the "correct" class, since we jump to the scope of the class when entering the external method declaration.

Reply:
This happens here: https://github.com/Fraunhofer-AISEC/cpg/blob/662a730234a9e16e6a69ae16caf30b06786226bf/cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/DeclarationHandler.kt#L151-L158

`holder` is either a class or a namespace.

Reply:
Ok, than it is better to go over the `currentClass`, I would still keep the currentMethod function though for future use.